File: leetcode/72_edit_distance.cpp
#include <string>
#include <utility>
#include <numeric>
#include <vector>
#include <algorithm>
#include <print>

class Solution {
public:
    int minDistance(std::string& word1, std::string& word2) {
        // Edge cases
        if (word1.empty()) {
            return word2.size();
        }
        if (word2.empty()) {
            return word1.size();
        }

        // DP problem, since by choosing the best of 3 operations on top of previously computed optimal subproblem gives the global best answer

        // Represent edit distance from word1[:i] -> word2[:j] as dp[i][j]
        // Then dp[i][j] = std::min({dp[i - 1][j] + 1, dp[i][j - 1] + 1, dp[i - 1][j - 1] + repl_cost})
        // where repl_cost = (word1[i - 1] == word2[j - 1]) ? 0 : 1

        // Since the computation of next state only requires values to left/top, and we only compute one row after its above row is completed, we can compress to a 1D dp array

        // Initialize base case values
        // Taking avantage of symmetry, let the shorter of the two words be in the row, and longer be in column
        const std::string* row_word = &word1; // pointer to row_word
        const std::string* col_word = &word2; // pointer to col_word
        if (row_word->size() > col_word->size()) { // swap pointer values / addresses
            std::swap(row_word, col_word);
        }

        std::vector<int> dp(row_word->size() + 1);
        std::iota(dp.begin(), dp.end(), 0);

        for (size_t i = 1; i <= col_word->size(); ++i) {
            int prev_diag = dp[0]; // dp[i - 1][j - 1]
            dp[0] = static_cast<int>(i); // dp[i][0] = i
            for (size_t j = 1; j <= row_word->size(); ++j) {
                int top = dp[j]; // dp[i - 1][j] is number of edits for col_word[:i-1] -> row_word[:j]; spend 1 delete operation for col_word[:i] -> row_word[:j]
                int left = dp[j - 1]; // dp[i][j - 1] is number of edits for col_word[:i] -> row_word[:j - 1]; spend 1 insert operation for col_word[:i] -> row_word[:j]
                int cost = ((*col_word)[i - 1] == (*row_word)[j - 1]) ? 0 : 1; // 1 if current chars don't match up; index i-1 and j-1 due to adding of extra row/col for DP matrix

                dp[j] = std::min({top + 1, left + 1, prev_diag + cost});
                prev_diag = top; // as j increments, dp[i - 1][j] -> dp[i - 1][j - 1]
            }
        }

        return dp.back();
    }
};

int main() {
    std::string word1 = "intention";
    std::string word2 = "execution";
    Solution s;
    std::print("{}", s.minDistance(word1, word2));
}

File: leetcode/72_edit_distance.py
class Solution1:
    def minDistance(self, word1: str, word2: str) -> int:
        # horse, ros -> 4x6 matrix
        # * h o r s e
        # r 1
        # o
        # s

        # Key idea 1: we simulate insertion/deletion by adjusting indices within a word
        # For example, ac -> abc comparison is moving j=1 to j=2 without moving i=1
        # Let us have a dp tracking min edits; dp will be (m + 1) x (n + 1)
        # dp[i][j] tracks min edits to transform word1[:i] into word2[:j]

        # update: dp[i][j] = dp[i - 1][j - 1] + 1 (i.e., all prev chars match, char at end of substring is edited)
        # insert into word2 / delete from word1: dp[i][j] = dp[i][j - 1] + 1
        # insert into word1 / delete from word2: dp[i][j] =  dp[i - 1][j] + 1
        # min of these 3 types of operations for each dp[i][j]

        m = len(word2)
        n = len(word1)
        dp = [[0 for _ in range(n + 1)] for _ in range(m + 1)]

        # dp[i][0] represents number of edits for word2 -> ""
        for i in range(1, m + 1):
            dp[i][0] = i
        # dp[0][j] represents number of edits for word1 -> ""
        for j in range(1, n + 1):
            dp[0][j] = j

        for i in range(1, m + 1): # word2
            for j in range(1, n + 1): # word1
                # Two cases; either we find a match and don't use an operation, or we need to use an operation (pick best)
                if word1[j - 1] == word2[i - 1]:
                    # match -> take num_edits as whatever was needed for word1[:j - 1] == word2[:i - 1]
                    dp[i][j] = dp[i - 1][j - 1]
                else:
                    best = min(dp[i - 1][j - 1], dp[i - 1][j], dp[i][j - 1])
                    dp[i][j] = best + 1

        return dp[-1][-1]


class Solution2:
    def minDistance(self, word1: str, word2: str) -> int:
        # Next, we optimize for space by noticing that dp only takes top left, left, and top cells
        # We can space-optimize this by using 1D dp using the length of the shorter word
        l1 = len(word1)
        l2 = len(word2)
        if l1 > l2:
            # swap so that word1 is always shorter / l1 <= l2
            l1, l2 = l2, l1
            word1, word2 = word2, word1

        dp = [i for i in range(l1 + 1)] # for word1="ros" this would be [0, 1, 2, 3]
        for i in range(1, l2 + 1):
            top_left = dp[0]
            dp[0] = i
            for j in range(1, l1 + 1):
                left = dp[j - 1]
                top = dp[j]
                best = min(left, top_left, top)
                tmp = dp[j] # store for next iteration
                dp[j] = best + 1 if word1[j - 1] != word2[i - 1] else top_left
                top_left = tmp

        return dp[-1]
